### ISSUE TITLE
Cabal install does not find locally installed dependency ("unknown package")

### ISSUE THREAD

**Issue by danse** [haskell/cabal#8714]
**Describe the bug**
I have package `b` depending on `a`. `a` is locally installed but when trying to build `b`, the dependency is not found. The error looks like:
```
Error: cabal: Could not resolve dependencies:
[__0] trying: b-0.1.0.0 (user goal)
[__1] unknown package: a (dependency of b)
[__1] fail (backjumping, conflict set: b, a)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: b, a
```

**To Reproduce**
Move to `a`, run `cabal install`, the installation succeeds. Now move to `b` depending on `a`, run `cabal install`, the installation fails with the error mentioned above. These are `v2-` commands.

**Expected behavior**
After installing `a`, i expected it to be locally visible for resolving dependencies.

**System information**
 - Ubuntu 20.04.5 LTS
 - `cabal` 3.8.1.0, `ghc` 8.6.5

**Additional context**
Executables provided by `a` are accessible and the library it provides gets correctly imported into `ghci`

**Comment by andreasabel** [comment]
> These are `v2-` commands.

In general, `v2` commands do not honor packages installed locally.  If you want your workflow, you have to use `v1` commands.

One way to make local packages available for `v2` is to add your own local package index: 
- https://cabal.readthedocs.io/en/latest/config.html#local-no-index-repositories

**Comment by gbaz** [comment]
This may be resolved in master by https://github.com/haskell/cabal/pull/8607#issuecomment-1322697152

**Comment by danse** [comment]
Thanks @andreasabel, probably i have a simplistic mental model of caching in cabal. I thought that once installed, packages kept no track of where they came from.

The doc you referenced seems a great solution to use version two for this use case. I tried adding:

```
repository a
        url: file+noindex:///home/name/a/dist-newstyle/sdist/
```

To `b.cabal`, since `install` leaves a `.tar.gz` for `a` in that dir, but it still does not find the package.

Thanks @gbaz for the reference to what is being merged. Is there a way to figure out which release will include those improvements, so that i can test again?

**Comment by andreasabel** [comment]
@danse worte
> adding ... To `b.cabal`

Ah, no, this has to go into `cabal.project` which you can put besides `b.cabal`.
`.cabal` files are for uploading to hackage, so they cannot mention things local to your machine/configuration.

**Comment by danse** [comment]
whoops, i thought that was the section about `.cabal` configuration. This works perfectly now, i just added a `packages: b.cabal`, to start the `cabal.project`. Thanks @andreasabel!

**Comment by andreasabel** [comment]
Yes, somehow the documentation should in section https://cabal.readthedocs.io/en/latest/config.html#repository-specification state explicitly that one can also specify repositories in the `cabal.project` file.

Also, information about repositories should be referenced from https://cabal.readthedocs.io/en/latest/cabal-project.html# e.g. in https://cabal.readthedocs.io/en/latest/cabal-project.html#cfg-field-active-repositories or more centrally.

**Comment by ulysses4ever** [comment]
> state explicitly that one can also specify repositories in the cabal.project file.

Sounds like am instance of #6319.

**Comment by josephcsible** [comment]
For reference, I asked this SO question a while back about how to do this: [How do I get cabal to use a local version of a package as a dependency for a Hackage package?](https://stackoverflow.com/q/69773999/7509065)
